### PR TITLE
Error for primary constructor in delegate definition

### LINKED ISSUE #8299: No syntax error for primary constructor in delegate definition
No error is reported for the following code:
```fsharp
type T(x: int) =
    delegate of int -> int
```

When sending this code to fsi, an error is produced:
```
error FS0193: internal error: Illegal definition for runtime implemented delegate method.
```

### PR DESCRIPTION
**Pull request by edgarfgp**
Attempt to fix #8299

### CODE REVIEW

File: tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// #Regression #Conformance #ObjectOrientedTypes #Delegates
```
Comment (vzarytovskii) [inline tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs:1]: We try not to add any new tests to the `fsharpqa`, I've mentioned about the ComponentTests in the https://github.com/dotnet/fsharp/issues/8299#issuecomment-1115849356, sorry for the delay.

Comment (edgarfgp) [reply tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs:1]: No worries :) . Thanks for taking the time . I will update this PR according with your suggestion

Comment (edgarfgp) [reply tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs]: @vzarytovskii I realised that the #8299 talks about to **show a syntax error**  for this rather that a compilation error. 

I’m a little confused the test that I added in this PR they succeeded I was expecting them to fail As if you send interactive it show an error 😢.

Comment (vzarytovskii) [reply tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs]: Yeah, when you call `compile` it will use FCS (i.e. compiler), and not fsi/interactive session.
Do you want to test compiler or fsi?

Comment (edgarfgp) [reply tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs]: > Yeah, when you call `compile` it will use FCS (i.e. compiler), and not fsi/interactive session.
> Did you intend to test compiler or fsi?


I think the expectation would to have one for **compiler**(that show a syntax error) and one for **fsi/interactive** session(that shows an error) . So they are both consistent .

Comment (vzarytovskii) [reply tests/fsharpqa/Source/Conformance/ObjectOrientedTypeDefinitions/DelegateTypes/E_InvalidDelegateDefinition.fs]: Yeah, so for compiler, you can simply leave this this test, and then fix the compiler itself, so the test passes (i.e. the compiler expected to fail).
For the FSI, you can use the `runFsi` function, here's the example of such test:
https://github.com/dotnet/fsharp/blob/d307f8d5091d9d1ac6e7e66948d4aac1596ab654/tests/FSharp.Compiler.ComponentTests/Scripting/Interactive.fs#L18-L29

Feel free to ping me on discord/slack if you have any questions.

**Comment by edgarfgp** [comment]
@vzarytovskii While working on this PR and asking in discord I wonder if there is Is anything preventing us to raise this during parsing and not in CheckDeclarations ?

**Comment by auduchinok** [comment]
@edgarfgp The parts that parse the constructor and the type body are implemented in different grammar rules and doing it in some parent rule would mean checking declarations produced by inner rules. The parser doesn't seem to be the right place to do this, as it's the job of the checker (that does a lot of similar checks already). Since this error is already reported for other kinds of functional F# types, the best place to implement reporting it for delegates would be somewhere near the existing cases.

**Comment by edgarfgp** [comment]
I have made the changes and wrote unit test and my test all passing locally but the one the CI gates is failing due to neg06.bsl differed: a bug or baseline may need updating

**Comment by vzarytovskii** [comment]
> I have made the changes and wrote unit test and my test all passing locally but the one the CI gates is failing due to neg06.bsl differed: a bug or baseline may need updating

Not yet sure why it's failing, gonna look at it later today or tomorrow morning.

**Comment by edgarfgp** [comment]
@vzarytovskii CI got even more angry with my attempt of updating the baseline tests 😅. I think I’m going to take couple days to try to understand this and then come back . I think I’m close to have this green

**Comment by vzarytovskii** [comment]
Oh, yeah, there was some infra issues.

File: src/Compiler/FSComp.txt [hunk 2]
```diff
@@ -1643,4 +1643,5 @@ reprStateMachineInvalidForm,"The state machine has an unexpected form"
 3520,invalidXmlDocPosition,"XML comment is not placed on a valid language element."
 3521,tcInvalidMemberDeclNameMissingOrHasParen,"Invalid member declaration. The name of the member is missing or has parentheses."
 3522,tcAnonRecdDuplicateFieldId,"The field '%s' appears multiple times in this record expression."
-3523,tcAnonRecdTypeDuplicateFieldId,"The field '%s' appears multiple times in this anonymous record type."
\ No newline at end of file
+3523,tcAnonRecdTypeDuplicateFieldId,"The field '%s' appears multiple times in this anonymous record type."
+3524,noArgumentsAreAllowedInDelegatePrimaryConstructor,"No argument are allowed in delegate primary constructor."
```
Comment (auduchinok) [inline src/Compiler/FSComp.txt:1647]: Perhaps we could reuse FS0552 here instead of adding another error number?

File: tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs.err.bsl [hunk 3]
```diff
@@ -0,0 +1 @@
+﻿invalid_delegate_definition.fs (1,6)-(1,15) No argument are allowed in delegate primary constructor.
```
Comment (auduchinok) [inline tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs.err.bsl:1]: @vzarytovskii Is it expected that the baseline doesn't contain the error numbers? Or is it checked in some other place?

I think it could be a good thing to check that numbers aren't changing during some changes in analysis code.

File: tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs [hunk 4]
```diff
@@ -0,0 +1,2 @@
+﻿type T(x: int) =
```
Comment (auduchinok) [inline tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs:1]: Please also add a test without any parameters, since an empty constructor is not allowed either.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 5]
```diff
@@ -3985,7 +3985,14 @@ module EstablishTypeDefinitionCores =
                 match fields' with 
                 | rf :: _ -> errorR (Error(FSComp.SR.tcInterfaceTypesAndDelegatesCannotContainFields(), rf.Range))
                 | _ -> ()
-
+            
+            let primaryConstructorInDelegateCheck(tyCon: Tycon, implicitCtorSynPats : SynSimplePats option) =
+                match implicitCtorSynPats with 
+                | None -> ()
+                | Some spats ->
+                    if tyCon.IsFSharpDelegateTycon then
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckDeclarations.fs:3993]: Is this function going to be called on non-delegate type definitions? If not, then this check is always true and is not needed.

**Review by auduchinok — COMMENTED** [review]
Just a few comments, @edgarfgp please take a look 🙂

File: src/Compiler/FSComp.txt (hunk 2 above)
Comment (edgarfgp) [reply src/Compiler/FSComp.txt:1647]: Humm. I could not find the  FS0552 on the list. if you can point me where is I can definitely reuse that :)

File: tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs (hunk 4 above)
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs]: Updated following your suggestion

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 5 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: Yeah that check will always be true . I have removed the  check

File: src/Compiler/FSComp.txt (hunk 2 above)
Comment (auduchinok) [reply src/Compiler/FSComp.txt:1647]: It's defined here:
https://github.com/dotnet/fsharp/blob/b5c1b447b3a6fadbae8e9d2f3012f0170bbec780/src/Compiler/FSComp.txt#L414

File: tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs.err.bsl (hunk 3 above)
Comment (vzarytovskii) [reply tests/FSharp.Compiler.ComponentTests/Conformance/DelegateTypes/invalid_delegate_definition.fs.err.bsl]: I think in these baselines it is, I will need to fix it.

File: DEVGUIDE.md [hunk 6]
```diff
@@ -128,8 +128,11 @@ If your changes involve modifying the list of language keywords in any way, (e.g
 ```shell
 dotnet build src\Compiler /t:UpdateXlf
 ```
+If you are on a Mac, you can run this command from the root of the repository:
 
-This only works on Windows/.NETStandard framework, so changing this from any other platform requires editing and syncing all of the XLF files manually.
+```shell
+sudo sh build.sh -c Release
```
Comment (vzarytovskii) [inline DEVGUIDE.md:134]: I think `sudo` is not required, since it's only updating files in the current directory?
```suggestion
sh build.sh -c Release
```

**Review by vzarytovskii — APPROVED** [review]

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 7]
```diff
@@ -3985,7 +3985,13 @@ module EstablishTypeDefinitionCores =
...
                 | rf :: _ -> errorR (Error(FSComp.SR.tcInterfaceTypesAndDelegatesCannotContainFields(), rf.Range))
                 | _ -> ()
-
+            
+            let primaryConstructorInDelegateCheck(implicitCtorSynPats : SynSimplePats option) =
+                match implicitCtorSynPats with 
+                | None -> ()
+                | Some spats ->
+                    let ctorArgNames, _ = TcSimplePatsOfUnknownType cenv true CheckCxs envinner tpenv spats
+                    if not ctorArgNames.IsEmpty then errorR (Error(FSComp.SR.parsOnlyClassCanTakeValueArguments(), m))
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckDeclarations.fs:3994]: I wonder, what is inside `ctorArgNames` when there is an empty constructor?

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:3994]:
```fsharp
namespace FSharpTest
    type T() =
        delegate of int -> int
```

Capture

**Review by auduchinok — APPROVED** [review]

**Comment by nojaf** [comment]
@edgarfgp congratulations on this PR!
A fine addition!

**Comment by vzarytovskii** [comment]
Thanks a lot for this @edgarfgp!

@KevinRansom @dsyme this LGTM, shall we get it in?

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
Looks good to me
